### PR TITLE
Add KL divergence

### PR DESCRIPTION
**Pull request by icantsayrural**
Equation:
<a href="https://www.codecogs.com/eqnedit.php?latex=D_{KL}(P||Q)&space;=&space;\sum_{i}&space;P(i)\log&space;\left(\frac{P(i)}{Q(i)}&space;\right&space;)" target="_blank"></a>

Requirements:
- Q can't have smaller support than P
- When P is 0, contribution to divergence is 0

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.vertices.ProposalVertex;
+
+public class KLDivergence {
+    public static double compute(ProposalVertex q, NetworkSamples samples) {
+        double divergence = 0;
+
+        for (int i = 0; i < samples.size(); i++) {
+            double pLogProb = samples.getLogOfMasterP(i);
+            double qLogProb = q.getLogOfMasterP(samples.getSample(i));
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: I think NetworkSamples should return something more structured than just a list of objects. Maybe a NetworkState that still has the ability to figure out which object maps to which vertex?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.junit.Test;
+
+import java.util.Collections;
+
+public class KLDivergenceTest {
+    @Test
+    public void testKLDivergence() {
+        GaussianVertex v1 = new GaussianVertex(0., 1.);
+        ConstantDoubleVertex v2 = new ConstantDoubleVertex(0.1);
+        AdditionVertex v3 = new AdditionVertex(v1, v2);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: `v3 = v1.plus(v2);`

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+        BayesianNetwork network = new BayesianNetwork(v3.getConnectedGraph());
+        NetworkSamples samples = MetropolisHastings
+            .withDefaultConfig()
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        ProposalVertex q = new GaussianVertex(0.1, 1.);
+
+        assertTrue(KLDivergence.compute(q, samples) < 3.0);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: This assertion is a bit arbitrary. Do we know what the KLDiv for these two distributions should be?

Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Maybe you can calculate the KLDiv for Gaussian(0.1, 1.0) and Gaussian(3.0, 1.0) and assert that the first is less than the second?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+            if (pLogProb != Double.NEGATIVE_INFINITY) {
+                if (qLogProb == Double.NEGATIVE_INFINITY) {
+                    throw new IllegalArgumentException("Q cannot have smaller support than P");
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: It might be worth including a link in the error message to a page that explains what "support" means. Unless we think that this is the kind of knowledge that all our users will have?

Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: You should capture this in a test case. Also the converse, i.e. it's OK for P to have a smaller support than Q.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import io.improbable.keanu.vertices.ProposalVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.AdditionVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+
+import java.util.Collections;
+
+public class KLDivergenceTest {
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Can you add a test that the divergence is zero when P == Q?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java [hunk 6]
```diff
@@ -21,7 +23,7 @@
 import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
 import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
 
-public class GaussianVertex extends DoubleVertex implements ProbabilisticDouble {
+public class GaussianVertex extends DoubleVertex implements ProbabilisticDouble, ProposalVertex {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java]: I like that the GaussianVertex can be used easily with KLDivergence but ideally this feature would be decoupled from the GaussianVertex itself. Can we add a `compute(ProbabilisticDouble q, NetworkSamples samples)` that does the wrapping inside the KLDivergence class?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/ProposalVertex.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package io.improbable.keanu.vertices;
+
+import java.util.List;
+
+public interface ProposalVertex {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/ProposalVertex.java]: Unless I've misunderstood something, it feels like it would be better to parametrise this class by `<T extends Tensor>`, and expect `List<T>` instead of `List<?>` (where the `T` will be the same type that the concrete Vertex class is parametrised on). 

That avoids the need for a yucky cast to `DoubleTensor` in your implementation of the method in `GaussianVertex`.

Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/ProposalVertex.java]: The proposal might not be a vertex. It might be a distribution represented by a bayesian network OR just a function that represents a distribution OR a single vertex that represents a distribution. Considering this interface is for KLDivergence and within the context of KL is very commonly called Q, maybe we could call it QDistribution and keep it in the same package as the KLDivergence class?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    @Test
+    public void testKLDivergence() {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: JUnit isn't great at helping developers describe the intention of their test but ideally the method name is the test name and the test name should describe what the assertion in the test is. This also encourages single assertions in tests which is good practice.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java [hunk 9]
```diff
@@ -130,4 +132,12 @@ public DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
             return DualNumber.createWithRespectToSelf(getId(), getValue());
         }
     }
+
+    @Override
+    public double getLogOfMasterP(List<?> values) {
+        if (values.size() != 1) {
+            throw new IllegalArgumentException("Univariate distributions takes exactly 1 value and you supplied: " + values.size());
+        }
+        return logProb((DoubleTensor) values.get(0));
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java]: You could replace this with a one-liner:

    return logProb((DoubleTensor) Iterables.getOnlyElement(values));

Which gives you a decent exception:

    java.lang.IllegalArgumentException: expected one element but was: <1.0, 1.0>

	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:322)
	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:294)
	at io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex.getLogOfMasterP(GaussianVertex.java:140)

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java [hunk 10]
```diff
@@ -130,4 +132,12 @@ public DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
             return DualNumber.createWithRespectToSelf(getId(), getValue());
         }
     }
+
+    @Override
+    public double getLogOfMasterP(List<?> values) {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java]: Could you add test coverage for the case when the list has size > 1?

**Comment by migwellian** [comment]
> "When P is 0, contribution to divergence is 0"

 - can you capture this in a test?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java (hunk 3 above)
Comment (icantsayrural) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Theoretically, the divergence should be 0. But we are sampling p here, so it'll just be arbitrarily low.

I like the idea of comparing the divergences between two different Gaussians!

Comment (icantsayrural) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Although, I see potential for flaky tests ... But I think we will be OK if the two qs are _very_ different.

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java (hunk 4 above)
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: Wikipedia has a section on support for all distributions, so I think we are OK

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java (hunk 9 above)
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/GaussianVertex.java]: Nice!

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        BayesianNetwork network = new BayesianNetwork(v3.getConnectedGraph());
+        NetworkSamples samples = MetropolisHastings
+            .withDefaultConfig()
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        ProbabilisticDouble q1 = new GaussianVertex(0.1, 1.);
+        ProbabilisticDouble q2 = new GaussianVertex(10.0, 1.);
+
+        assertTrue(KLDivergence.compute(q1, samples) < KLDivergence.compute(q2, samples));
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: It's a good idea to use Hamcrest matchers for asserts. The code reads more fluently and the error messages are much more informative.

For example:

    assertThat(KLDivergence.compute(q1, samples), lessThan(KLDivergence.compute(q2, samples)));

yields messages like:

    java.lang.AssertionError: 
    Expected: a value less than <14325.522530625412>
         but: <1435844.110279129> was greater than <14325.522530625412>

Whereas the `assertTrue` statement only tells you that there's an `java.lang.AssertionError`, with no message.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class KLDivergenceTest {
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Nice tests! They have good names that tell you exactly what they're testing for.

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/QDistribution.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package io.improbable.keanu.algorithms.variational;
+
+import io.improbable.keanu.network.NetworkState;
+
+public interface QDistribution {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/QDistribution.java]: There are no instantiations of this interface in the code base, except for your TestQDistribution which only returns a constant. Can we add some more real-world examples, and tests to accompany them?

Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/QDistribution.java]: Can you add some JavaDoc that explains what a Q Distribution is?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
+package io.improbable.keanu.algorithms.variational;
+
+import com.google.common.collect.Iterables;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.network.NetworkState;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.dbl.probabilistic.ProbabilisticDouble;
+
+public class KLDivergence {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: 👍 It's a very nice use of a lambda actually. It's a single lambda used effectively as a reduce function. Ace.

[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: I actually like the idea of using the `QDistribution` interface here so we begin to see it in the code. What're your thoughts on something like:

```
    public static double compute(ProbabilisticDouble q, NetworkSamples samples) {
        return compute(samples, i -> new ProbabilisticDoubleAsQDistribution(q).getLogOfMasterP(samples.getNetworkState(i)));
    }
```

```
    private static class ProbabilisticDoubleAsQDistribution implements QDistribution {

        final ProbabilisticDouble q;

        ProbabilisticDoubleAsQDistribution(ProbabilisticDouble q) {
            this.q = q;
        }

        @Override
        public double getLogOfMasterP(NetworkState state) {
            DoubleTensor vertexValue = state.get(Iterables.getOnlyElement(state.getVertexIds()));
            return q.logProb(vertexValue);
        }
    }
```

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+    private NetworkSamples createNetworkSamplesWithOneVertexAndOneSample(double p) {
+        Map<VertexId, List<DoubleTensor>> samplesByVertex = new HashMap<>();
+        samplesByVertex.put(new VertexId(), Collections.singletonList(DoubleTensor.scalar(1.)));
+        List<Double> logOfMasterPForEachSample = Collections.singletonList(p);
+
+        return new NetworkSamples(samplesByVertex, logOfMasterPForEachSample, 1);
+    }
+
+    private class TestGaussianQDistribution extends GaussianVertex implements QDistribution {
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: Nice

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        QDistribution qDist = mock(QDistribution.class);
+        when(qDist.getLogOfMasterP(any(NetworkState.class))).thenReturn(identicalProb);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(identicalProb);
+
+        assertEquals(0., KLDivergence.compute(qDist, samples), 1e-6);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfPProbIsNotZeroButQIs() {
+        QDistribution q = mock(QDistribution.class);
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: you can pull out `q` as a field of the class

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Math.log(0.5));
+
+        KLDivergence.compute(q, samples);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPIsZeroButQIsNot() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Math.log(0.5));
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: you could pull this out into a method:

    private QDistribution mockThatReturnsConstantValue(double identicalProb) {
        QDistribution q = mock(QDistribution.class);
        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(identicalProb);
        return q;
    }

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        QDistribution q1 = new TestGaussianQDistribution(0.1, 1.);
+        QDistribution q2 = new TestGaussianQDistribution(10.0, 1.);
+
+        assertThat(KLDivergence.compute(q1, samples), lessThan(KLDivergence.compute(q2, samples)));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfNetworkStateHasMoreThanOneVertexAndQIsProbabilisticDouble(){
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: I think this could do with some explanation in a javadoc comment. Why is this a bad thing to do? Is this just something that's not yet supported, or is it a logical mistake on the user's part? If the latter - what should the user do instead?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+        for (int i = 0; i < samples.size(); i++) {
+            double pLogProb = samples.getLogOfMasterP(i);
+
+            NetworkState state = samples.getNetworkState(i);
+            double qLogProb = qLogProbCalculator.apply(state);
+
+            if (pLogProb != Double.NEGATIVE_INFINITY) {
+                if (qLogProb == Double.NEGATIVE_INFINITY) {
+                    throw new IllegalArgumentException("Q cannot have smaller support than P");
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: not tested

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java (hunk 18 above)
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java]: A neat library you can use here is https://junit.org/junit4/javadoc/latest/org/junit/rules/ExpectedException.html#expectMessage(java.lang.String)

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java (hunk 19 above)
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java]: It is, but as discussed I'll use the library to expect a message to be more clear in my tests.
